Feature seaborn like api

ChandraDevs1 opened the pull request:
feat: Add seaborn-like API functions for plotting

- Implemented `plot_chromatogram`, `plot_spectrum`, `plot_mobilogram`, and `plot_peakmap` functions.
- Exposed functions in `__init__.py` for easy access.
- Functions wrap existing `data.plot()` functionality with a simplified API.

resolves : #51 

@jcharkow Please review the PR and let me know if you have any suggestions.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced new visualization capabilities for chromatogram, spectrum, mobilogram, and peakmap data, enhancing the range of data analysis tools available to users.

- **Documentation**
  - Updated the API documentation to include a new section on Plotly-based visualization extensions, offering clearer guidance on these new features.

- **Tests**
  - Expanded the test suite to validate the new plotting functions and improve error handling for invalid or missing data.
  
- **Chores**
  - Improved the development environment by configuring a non-interactive plotting backend and updating version control to ignore system-specific files.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1500,19 @@ def _get_plot_backend(backend: str | None = None):
     _backends[backend_str] = module
     return module
+
+def plot_chromatogram(data, x=None, y=None, backend=None, **kwargs):
```
x, y should not be optional arguments and therefore should not be assigned None

backend will also crash if it is None but you might be able to get away with keeping this none. 
What you can do is if backend is not set, check if matplotlib is installed (try import statement) and if so then use "ms_matplotlib" as the backend, then try bokeh, then try plotly. If no backends are installed then throw an exception.

ChandraDevs1 replied:
done

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1500,19 @@ def _get_plot_backend(backend: str | None = None):
     return module
+
+def plot_chromatogram(data, x=None, y=None, backend=None, **kwargs):
+    """Plot a chromatogram using a seaborn-like API."""
```
Documentaiton can be improved, please use google style doc string.

ChandraDevs1 replied:
done

jcharkow reviewed (changes requested):
Good start! Please see comments. Also please add tests for these functions. You can extend  test/plot_chromatogram.py, test/plot_spectrum.py etc. with the new functions you added. You do not need to test all the different parameters but it would be good to test edge cases e.g. if y is not given or backend is not given.

ChandraDevs1 commented on pyopenms_viz/_core.py:
```diff
@@ -13,7 +13,8 @@
 import re
-
+from typing import Optional
+import pandas as pd
```
done.

ChandraDevs1 commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1500,177 @@ def _get_plot_backend(backend: str | None = None):
+    """
+    if x is None or y is None:
+        raise ValueError("Both `x` and `y` must be provided.")
+    return data.plot(x=x, y=y, z=z, kind="peakmap", backend=backend, **kwargs)
```
done

ChandraDevs1 reviewed (commented):
resolves all the comments

ChandraDevs1 commented on pyopenms_viz/__init__.py:
```diff
@@ -196,5 +201,10 @@ def _get_plot_backend(backend: str | None = None):
+    'plot_spectrum',
+    'plot_mobilogram',
+    'plot_peakmap'
+]
```
done

ChandraDevs1 replied:
right

ChandraDevs1 commented:
@jcharkow  I have resolves the comments Please Review the PR and let me know If changes were required.
thanking you

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1499,186 @@ def _get_plot_backend(backend: str | None = None):
+        The column name for the x-axis data.
+    y : str
+        The column name for the y-axis data.
+    backend : str, optional
+        The backend to use for plotting. If not provided, it will try to use
+        matplotlib, bokeh, or plotly in that order.
```
As I said before, this is not currently implemented. Please implement or make backend argument mandatory.

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1499,186 @@ def _get_plot_backend(backend: str | None = None):
+    >>> data = pd.DataFrame({'im': [1, 2, 3], 'intensity': [10, 20, 30]})
+    >>> oms_viz.plot_mobilogram(data, x='im', y='intensity', backend='ms_plotly')
+    """
+    if x is None or y is None:
+        raise ValueError("Both `x` and `y` must be provided.")
```
This is not relevant anymore because x and y are mandatory arguments.

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1499,186 @@ def _get_plot_backend(backend: str | None = None):
+
+    Raises
+    ------
+    ValueError
+        If `x` or `y` is not provided.
```
Edit documentation when valueError checking is removed

ChandraDevs1 replied:
done

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1499,186 @@ def _get_plot_backend(backend: str | None = None):
+        raise ValueError("Both `x` and `y` must be provided.")
+    if z is None:
+        raise ValueError("`z` must be provided for peakmap visualization.")
+    if data.empty:
+        raise ValueError("DataFrame is empty, cannot plot with no data.")
```
This is a good check, add this to methods above and update documentation accordingly.

jcharkow reviewed (changes requested):
Good progress, please see comments above.

jcharkow commented:
Also please merge the main branch into this branch so that the new tests can be triggered via github actions.

ChandraDevs1 commented:
@jcharkow Please review the PR
\

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1500,191 @@ def _get_plot_backend(backend: str | None = None):
+        - If `x` or `y` contain non-numeric data
+    """
+    # Validate input data
+    if data.empty:
+        raise ValueError("DataFrame is empty, cannot plot with no data.")
+    
+    if x not in data.columns:
+        raise ValueError(f"Column '{x}' not found in DataFrame")
+    if y not in data.columns:
+        raise ValueError(f"Column '{y}' not found in DataFrame")
+
+    if not pd.api.types.is_numeric_dtype(data[x]):
+        raise ValueError(f"Column '{x}' must contain numeric data")
+    if not pd.api.types.is_numeric_dtype(data[y]):
+        raise ValueError(f"Column '{y}' must contain numeric data")
```
These are all good checks however most are already checked for in _core.py so they are not needed here.

However data.empty is not checked and it a good check. Please delete that here and add that check to _core.py  BasePlot class under __init__ (e.g. under _verify_column)

ChandraDevs1 replied:
done

jcharkow commented on pyopenms_viz/_core.py:
```diff
@@ -1500,3 +1500,191 @@ def _get_plot_backend(backend: str | None = None):
+        - If `x` or `y` columns are missing
+        - If `x` or `y` contain non-numeric data
+    """
+    if data.empty:
+        raise ValueError("DataFrame is empty, cannot plot with no data.")
+    
+    if x not in data.columns:
+        raise ValueError(f"Column '{x}' not found in DataFrame")
+    if y not in data.columns:
+        raise ValueError(f"Column '{y}' not found in DataFrame")
+
+    if not pd.api.types.is_numeric_dtype(data[x]):
+        raise ValueError(f"Column '{x}' must contain numeric data")
+    if not pd.api.types.is_numeric_dtype(data[y]):
+        raise ValueError(f"Column '{y}' must contain numeric data")
+
+    return data.plot(x=x, y=y, kind="spectrum", backend=backend, **kwargs)
```
see comments above and apply to classes below as well.

ChandraDevs1 replied:
done

jcharkow reviewed (changes requested):
Coming alone. A few more minor comments. Also I've added the new tests github actions workflow you have added to this PR and it seems that the new tests that you added are failing so please address this: 
https://github.com/OpenMS/pyopenms_viz/actions/runs/13831508392/job/38729655840?pr=57

ChandraDevs1 commented:
I have completely passed the CI testcases and also made improvements as your suggestion. So kindly @t0mdavid-m @jcharkow Please review the PR. and let me now further changes or updates ..

thanking you for having the time to review  it .

ChandraDevs1 commented:
@jcharkow, @t0mdavid-m, I have successfully implemented all the required changes based on your valuable suggestions, test failures, and Coberabbit AI recommendations. I would appreciate it if you could take some time to review the PR and let me know if any further updates are needed. Thank you!

jcharkow commented on test/test_spectrum.py:
```diff
@@ -1,11 +1,11 @@
 """
-test/test_spectrum
+test/plotting/test_spectrum
```
keep as test/test_spectrum since this is the file path

jcharkow commented on test/test_peakmap.py:
```diff
@@ -1,11 +1,11 @@
 """
-test/test_peakmap
+test/plotting/test_peakmap
```
revert back to test/test_peakmap as this is the filepath

jcharkow commented on test/test_peakmap.py:
```diff
@@ -18,6 +18,8 @@
     ],
 )
 def test_peakmap_plot(featureMap_data, snapshot, kwargs):
+    if featureMap_data.empty:
+        pytest.skip("DataFrame is empty, skipping plot.")
```
This line (and in functions after not need this as by definition this should always be present)

jcharkow reviewed (changes requested):
Almost there! Just need to remove some unneeded files and address question with tests

ChandraDevs1 commented:
@jcharkow, as you mentioned, I have removed the suggested unneeded files and addressed the questions with tests.
and please let me for further updates .

thanks

jcharkow commented on test/test_peakmap.py:
```diff
@@ -32,6 +33,8 @@ def test_peakmap_plot(featureMap_data, snapshot, kwargs):
 
 # use peakmap to plot a 2D spectrum
 def test_peakmap_mz_im(featureMap_data, snapshot):
+    if featureMap_data.empty:
+        pytest.skip("DataFrame is empty, skipping plot.")
```
Please remove this and all instances like this across the test files that were added.

jcharkow reviewed (changes requested):
Just remove these `pytest.skip` lines across the different tests and provided everything passes should be good to merge.

ChandraDevs1 commented:
@jcharkow, thank you so much for your time and effort in patiently reviewing such a large PR. That means a lot! Please review it once more and let me know if any further updates are needed.

jcharkow reviewed (approved):
Looks good! Ready to merge
